File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    answer = []
    valid = [0] * 26
    
    for term in terms:
        a, t = term.split(" ")
        valid[ord(a)-ord('A')] = int(t)
    t_y, t_m, t_d = today.split(".")
    t_y = int(t_y)
    t_m = int(t_m)
    t_d = int(t_d)
    t_total = 336*t_y + 28*t_m + t_d
    for i, privacy in enumerate(privacies):
        day, alpha = privacy.split(" ")
        y, m, d = day.split(".")
        y = int(y)
        m = int(m)
        d = int(d)
        if d == 1:
            d = 28
            m += valid[ord(alpha) - ord('A')] - 1
        else:
            d -= 1
            m += valid[ord(alpha) - ord('A')]
        if (m > 12):
            m -= 12
            y += 1
        total = 336*y+28*m+d
        if t_total > total:
            answer.append(i+1)
    return answer

File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;

class Solution {
    public int[] solution(String today, String[] terms, String[] privacies) {
        List<Integer> list = new ArrayList<>();
        HashMap<String, String> map = new HashMap<>();
        
        for (String term : terms) {
            map.put(term.split(" ")[0], term.split(" ")[1]);
        }
        
        int year = Integer.parseInt(today.split("\\.")[0]);
        int month = Integer.parseInt(today.split("\\.")[1]);
        int day = Integer.parseInt(today.split("\\.")[2]);
        
        for (int i = 0; i < privacies.length; i++) {
            String date = privacies[i].split(" ")[0];
            int type = Integer.parseInt(map.get(privacies[i].split(" ")[1]))*28;
            int num = (year - Integer.parseInt(date.split("\\.")[0]))*28*12
                + (month - Integer.parseInt(date.split("\\.")[1]))*28
                + (day - Integer.parseInt(date.split("\\.")[2]));
            
            if (num >= type) {
                list.add(i + 1);
            }
        }
        
        return list.stream().filter(i -> i != null).mapToInt(i -> i).toArray();
    }
}